core/domain/exp_services.py

```diff
@@ -1997,6 +2028,15 @@ def get_exp_with_draft_applied(exp_id, user_id):
     exploration = get_exploration_by_id(exp_id)
     if exp_user_data:
         if exp_user_data.draft_change_list:
+            if (
+                    exploration.version >
+                    exp_user_data.draft_change_list_exp_version):
+                logging.info(
+                    'Exploration and draft versions out of sync, trying '
+                    'to upgrade draft version to match exploration\'s.')
+                try_upgrading_draft_to_exp_version(exp_id, user_id)
```

Comment:
I'm a little iffy about this, because this is a "get" method but this command is basically doing a put() internally.

I think instead it's better to pass the draft change list and attempt an upgrade without saving, and then apply the upgraded change list if possible to the exploration. We would only save the upgraded change list into the datastore at the time we would normally be doing a PUT operation on the change list. (This is similar to how we handle exploration schema upgrades -- we don't put() the upgraded exploration if the context is a GET.)

Reply:
Ahh, I think that's totally fair. 

For drafts, it seems that we can just remove this part. The new draft will get automatically saved in EditorAutosaveHandler.

```
class EditorAutosaveHandler(ExplorationHandler):
    def put(self, exploration_id):
        try:
            change_list_dict = self.payload.get('change_list')
            change_list = [
                exp_domain.ExplorationChange(change)
                for change in change_list_dict]
            version = self.payload.get('version')
            exp_services.create_or_update_draft(
                exploration_id, self.user_id, change_list, version,
                datetime.datetime.utcnow())
    
```

Reply:
I realized we check exp_user_data in places like 
```
def get_user_exploration_data(
        user_id, exploration_id, apply_draft=False, version=None):
# ...
    is_valid_draft_version = (
        is_version_of_draft_valid(
            exploration_id, exp_user_data.draft_change_list_exp_version)
        if exp_user_data and exp_user_data.draft_change_list_exp_version
        else None)
# ...
```
I'll refactored this method a little to get rid of the additional check and instead read from the return result from get_exp_with_draft_applied

Reply:
Adopted Sean's suggestion here to have get_exp_with_draft_applied return None if failed to apply draft.